package elb;

enum MsgTypeId {
    GetHostReqId         = 1; //api get a host from agent
    GetHostRspId         = 2; //agent give a host to api
    ReportReqId          = 3; //api report call result to agent
    GetRouteByToolReqId  = 4; //tool get route from agent
    GetRouteByToolRspId  = 5; //agent give route to tool
    GetRouteByAgentReqId = 6; //agent get route from dnsserver
    GetRouteByAgentRspId = 7; //dnsserver give route to agent
    ReportStatusReqId    = 8; //agent report lb status to reporter
    CacheGetRouteReqId   = 9; //为支持API cache: api向agent发起获取\更新路由请求
    CacheGetRouteRspId   = 10;//为支持API cache: agent给api返回的路由应答
    CacheBatchRptReqId   = 11;//为支持API cache: api向agent批量上报若干成功结果
}

//represent a remote node
message HostAddr {
    required int32 ip  = 1;
    required int32 port = 2;
}

//api get a host from agent (UDP)
message GetHostReq {
    required uint32 seq   = 1;
    required int32 modid  = 2;
    required int32 cmdid  = 3;
}

//agent give a host to api (UDP)
message GetHostRsp {
    required uint32 seq    = 1;
    required int32 modid   = 2;
    required int32 cmdid   = 3;
    required int32 retcode = 4;
    optional HostAddr host = 5;
}

//api report call result to agent
message ReportReq {
    required int32 modid   = 1;
    required int32 cmdid   = 2;
    required HostAddr host = 3;
    required int32 retcode = 4;
    optional uint32 tcost  = 5;//如果调用失败，需要收集调用用时
}

//agent get route from dnsserver (TCP), or tool get route from agent (UDP)
message GetRouteReq {
    required int32 modid = 1;
    required int32 cmdid = 2;
}

//dnsserver give route to agent (TCP), or agent give route to tool (UDP)
message GetRouteRsp {
    required int32 modid    = 1;
    required int32 cmdid    = 2;
    repeated HostAddr hosts = 3;
}

//host call result
message HostCallResult {
    required int32 ip      = 1;
    required int32 port    = 2;
    required uint32 succ   = 3;
    required uint32 err    = 4;
    required bool overload = 5;
}

//agent report lb status to reporter
message ReportStatusReq {
    required int32 modid            = 1;
    required int32 cmdid            = 2;
    required int32 caller           = 3;
    repeated HostCallResult results = 4;
    required uint32 ts              = 5;//sec
}

//为支持API cache: api向agent发起获取\更新路由请求
message CacheGetRouteReq {
    required int32 modid   = 1;
    required int32 cmdid   = 2;
    required int64 version = 3;//首次获取时，version=-1
}

//为支持API cache: agent给api返回的路由应答
message CacheGetRouteRsp {
    required int32 modid    = 1;
    required int32 cmdid    = 2;
    required int64 version  = 3;//agent不存在此mod，则version=-1
    optional bool overload  = 4;
    repeated HostAddr route = 5;
}

//host batch call result
message HostBatchCallRes {
    required int32 ip       = 1;
    required int32 port     = 2;
    required uint32 succCnt = 3;
}

//为支持API cache: api向agent批量上报若干成功结果
message CacheBatchRptReq {
    required int32 modid              = 1;
    required int32 cmdid              = 2;
    repeated HostBatchCallRes results = 3;
}
